### PR TITLE
[fx] Add concrete info prop

### PR DESCRIPTION
**Pull request by Cypher30**
# What's New?

In this PR, I provide `ConcreteInfoProp` to facilitate the meta info estimation, we also provide the time profiler that track the running time of each node. You could check the docstring of function `_profile_concrete` for more information.

For example, now you could run
```python
import torch
from torch.fx import symbolic_trace
from colossalai.fx.passes import ConcreteInfoProp

BATCH_SIZE = 2
DIM_IN = 4
DIM_HIDDEN = 16
DIM_OUT = 16
model = torch.nn.Sequential(
    torch.nn.Linear(DIM_IN, DIM_HIDDEN), 
    torch.nn.Linear(DIM_HIDDEN, DIM_OUT),
    ).cuda()
input_sample = torch.rand(BATCH_SIZE, DIM_IN, device="cuda")
gm = symbolic_trace(model)
interp = ConcreteInfoProp(gm)
interp.run(input_sample)
print(interp.summary(unit='kb')) 
```

and get the following results
```bash
 Op type       Op             Forward time             Backward time    SAVE_FWD_IN    FWD_OUT    FWD_TMP    BWD_OUT    BWD_TMP
-----------  -------  -----------------------  ------------------------  -------------  ---------  ---------  ---------  ---------
placeholder  input_1                    0.0 s                     0.0 s          False    0.00 KB    0.00 KB    0.00 KB    0.00 KB
call_module       _0  0.0003993511199951172 s     0.00706791877746582 s          False    0.50 KB    0.00 KB    0.03 KB    0.66 KB
call_module       _1   6.29425048828125e-05 s  0.00018286705017089844 s          False    0.50 KB    0.00 KB    0.12 KB    0.81 KB
     output   output                    0.0 s                     0.0 s           True    0.00 KB    0.00 KB    0.00 KB    0.00 KB
```

### CODE REVIEW

File: colossalai/fx/profiler/profiler.py [hunk 1]
```diff
@@ -27,33 +28,104 @@ def is_autogradable(x):
...
+        x = x.detach()
+        x.requires_grad = requires_grad
+
+    return x
 
 
 def _profile_concrete(target: Callable, *args, **kwargs) -> Tuple[Tuple[Any, ...], GraphInfo]:
-    """
-    Profile a Callable function with args and kwargs on concrete devices.
+    """Profile a Callable function with args and kwargs on concrete devices by https://github.com/Cypher30
```
Comment (Cypher30) [inline colossalai/fx/profiler/profiler.py:41]: Here is the shortcut for you guys to check what I'm doing!!

File: colossalai/fx/passes/concrete_info_prop.py [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+    Extract a TensorMetadata NamedTuple describing `result`.
+    """
+    shape = result.shape
+    dtype = result.dtype
+    requires_grad = result.requires_grad
+    stride = result.stride()
+    numel = result.numel()
+    is_tensor = True
+
+    return TensorMetadata(shape, dtype, requires_grad, stride, numel, is_tensor)
```
Comment (super-dainiu) [inline colossalai/fx/passes/concrete_info_prop.py:38]: You don't need these for concrete_info_prop

File: colossalai/fx/passes/concrete_info_prop.py [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+        result, meta_info = super().run_node(n)
+
+        def extract_tensor_meta(obj):
+            if isinstance(obj, torch.Tensor):
+                return _extract_tensor_metadata(obj)
+            else:
+                return TensorMetadata(None, None, False, None, 0, False)
+
+        tensor_meta = tree_map(extract_tensor_meta, result)
+        n.meta['tensor_meta'] = tensor_meta
```
Comment (super-dainiu) [inline colossalai/fx/passes/concrete_info_prop.py:102]: don't need this

File: colossalai/fx/passes/concrete_info_prop.py [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+                details on semantics
+            args (Tuple): Tuple of positional args for this invocation
+            kwargs (Dict): Dict of keyword arguments for this invocation
+
+        Return
+            result (Any): The argument value that was retrieved
+            meta_info (MetaInfo): The memory cost and forward & backward time.
+        """
+        assert not isinstance(target, str)
+        return profile_function(target, "cuda:0")(*args, **kwargs)
```
Comment (super-dainiu) [inline colossalai/fx/passes/concrete_info_prop.py:172]: Should support `cpu` and `cuda:1`?

File: colossalai/fx/passes/concrete_info_prop.py (hunk 3 above)
Comment (Cypher30) [reply colossalai/fx/passes/concrete_info_prop.py:102]: okay

File: colossalai/fx/passes/concrete_info_prop.py (hunk 2 above)
Comment (Cypher30) [reply colossalai/fx/passes/concrete_info_prop.py:38]: okay

File: colossalai/fx/passes/concrete_info_prop.py (hunk 4 above)
Comment (super-dainiu) [reply colossalai/fx/passes/concrete_info_prop.py:172]: nevermind `cpu` is not a good option

Comment (Cypher30) [reply colossalai/fx/passes/concrete_info_prop.py]: I've fixed this by customized run function for ConcreteInfoProp, finding the device in the args and store it in `self.device`

File: colossalai/fx/profiler/profiler.py [hunk 5]
```diff
@@ -257,18 +331,27 @@ def f(*args: Tuple[Argument, ...], **kwargs: Dict[str, Any]) -> Any:
         # If there is an argument that this `call_module` is inplace, we should
         # still run the profiling but discard some results regarding `module`.
         inplace = getattr(module, 'inplace', False)
+
+        # calculate parameter size
+        param_size = 0
+        for param in module.parameters():
+            param_size += param.numel() * param.element_size()
```
Comment (super-dainiu) [inline colossalai/fx/profiler/profiler.py:338]: can use the `parameter_size` defined in `memory.py`

Comment (Cypher30) [reply colossalai/fx/profiler/profiler.py:338]: okay

Comment (Cypher30) [reply colossalai/fx/profiler/profiler.py]: fixed~

File: colossalai/fx/profiler/profiler.py [hunk 6]
```diff
@@ -27,33 +28,104 @@ def is_autogradable(x):
...
-# x.detach() will change the unique identifier of data_ptr
-# we need to handle this in a stupid way
-def detach(x):
+def detach_variables(x):
     if isinstance(x, torch.Tensor):
         requires_grad = x.requires_grad
-        x.requires_grad_(False)
-        x.requires_grad_(requires_grad)
+        x = x.detach()
+        x.requires_grad = requires_grad
```
Comment (super-dainiu) [inline colossalai/fx/profiler/profiler.py:35]: wait, this is incorrect for `MetaInfoProp`, please recover the comment and the original function

File: colossalai/fx/profiler/profiler.py [hunk 7]
```diff
@@ -257,18 +331,25 @@ def f(*args: Tuple[Argument, ...], **kwargs: Dict[str, Any]) -> Any:
...
+            # currently we set the fwd_mem_tmp of ReLU to zero
+            if type(module) in [torch.nn.modules.activation.ReLU]:
                 meta.save_fwd_in = False
                 meta.bwd_mem_out = 0
+                meta.fwd_mem_tmp = 0
+        else:
+            out, meta = _profile_concrete(func, *args, **kwargs)
+
+        # grad for param will not be counted
+        meta.bwd_mem_tmp -= param_size
```
Comment (super-dainiu) [inline colossalai/fx/profiler/profiler.py:352]: a hypothesis here: why bwd_temp captures the grad of weights in meta profiling?

Comment (Cypher30) [reply colossalai/fx/profiler/profiler.py]: fixed~

File: colossalai/fx/profiler/profiler.py [hunk 8]
```diff
@@ -206,12 +285,26 @@ def f(*args: Tuple[Argument, ...], **kwargs: Dict[str, Any]) -> Any:
...
                 meta.save_fwd_in = False
                 meta.bwd_mem_out = 0
+                meta.fwd_mem_tmp = 0
+        else:
+            out, meta = _profile_concrete(func, *args, **kwargs)
+
+        # find the grad for parameter in args and kwargs
+        param_size = 0
+
+        def get_param_size(x):
```
Comment (Cypher30) [inline colossalai/fx/profiler/profiler.py:300]: this part is for retrieving param size in args and kwargs

**Review by super-dainiu — APPROVED** [review]

**Pull request merged**